### PR TITLE
KTOR-5225 Add callbacks to save application state for OAuth2

### PR DESCRIPTION
**Pull request by rsinukov**
https://youtrack.jetbrains.com/issue/KTOR-5225

### CODE REVIEW

File: ktor-server/ktor-server-plugins/ktor-server-auth/jvm/src/io/ktor/server/auth/OAuth.kt [hunk 1]
```diff
@@ -161,15 +196,31 @@ public sealed class OAuthAccessTokenResponse : Principal {
...
         val accessToken: String,
         val tokenType: String,
         val expiresIn: Long,
         val refreshToken: String?,
-        val extraParameters: Parameters = Parameters.Empty
-    ) : OAuthAccessTokenResponse()
+        val extraParameters: Parameters = Parameters.Empty,
+    ) : OAuthAccessTokenResponse() {
+
+        public var state: String? = null
```
Comment (whyoleg) [inline ktor-server/ktor-server-plugins/ktor-server-auth/jvm/src/io/ktor/server/auth/OAuth.kt:210]: just a question: is it expected, that while `OAuth2` is `data class` this **new** property will not appear in `copy`/`componentN`/`equals`/`hashCode`?

File: ktor-server/ktor-server-plugins/ktor-server-auth/jvm/src/io/ktor/server/auth/OAuth.kt [hunk 2]
```diff
@@ -117,6 +118,40 @@ public sealed class OAuthServerSettings(public val name: String, public val vers
             emptyList(),
             accessTokenInterceptor
         )
+
+        @Deprecated("This constructor will be removed", level = DeprecationLevel.HIDDEN)
```
Comment (whyoleg) [inline ktor-server/ktor-server-plugins/ktor-server-auth/jvm/src/io/ktor/server/auth/OAuth.kt:122]: minor: "This constructor will be removed" -> "Binary compatibility with 2.x"?

**Review by e5l — APPROVED** [review]
lgtm

File: ktor-server/ktor-server-plugins/ktor-server-auth/jvm/src/io/ktor/server/auth/OAuth.kt (hunk 1 above)
Comment (rsinukov) [reply ktor-server/ktor-server-plugins/ktor-server-auth/jvm/src/io/ktor/server/auth/OAuth.kt:210]: It's really difficult to add it without breaking binary compatibility. In practice, I don't expect users to do `copy`/`componentN` on principals. And for equals and hashcode, having `accessToken` there should be enough.

File: ktor-server/ktor-server-plugins/ktor-server-auth/jvm/src/io/ktor/server/auth/OAuth.kt (hunk 2 above)
Comment (rsinukov) [reply ktor-server/ktor-server-plugins/ktor-server-auth/jvm/src/io/ktor/server/auth/OAuth.kt:122]: IMO, deprecation messages should be for users explaining why this is deprecated, not for us explaining why it is kept.

Comment (whyoleg) [reply ktor-server/ktor-server-plugins/ktor-server-auth/jvm/src/io/ktor/server/auth/OAuth.kt:122]: AFAIK, declarations with `level=HIDDEN` will be not visible to user, until they open source code. They even should feel this change at all, as new parameter has default value, but ok.

Comment (rsinukov) [reply ktor-server/ktor-server-plugins/ktor-server-auth/jvm/src/io/ktor/server/auth/OAuth.kt:122]: Yes, but for consistency, I prefer it this way.

**Pull request merged**